app/src/main/java/pl/edu/pjwstk/s999844/shoppinglist/activities/MainActivity.kt

```diff
@@ -85,21 +86,44 @@ class MainActivity : AbstractShoppingActivity() {
 	@Suppress("UNUSED_PARAMETER")
 	fun onClickFloatingButton(view: View) = startActivity(Intent(baseContext, AddItemActivity::class.java))
 
-	private fun changeItemCallback(item: RequiredItem, change: Int) {
+	/**
+	 * Update item's amount.
+	 *
+	 * @param[item] what to change amount of
+	 * @param[change] how much it should change, `null` for deleting it
+	 */
+	private fun changeItemCallback(item: RequiredItem, change: Int?) {
 		val dbItem: RequiredItem = shoppingListDao.findById(item.id)
 			?: return
 
-		val newAmount: Int = dbItem.amount + change
-		if (newAmount <= 0) {
+		if (change == null) {
 			shoppingListDao.delete(dbItem)
 		} else {
-			dbItem.amount = newAmount
+			dbItem.amount = change.plus(dbItem.amount).coerceAtLeast(0)
 			shoppingListDao.update(dbItem)
 		}
 	}
 
 	private fun observeDatabaseChange(items: List<RequiredItem>) {
-		(binding.mainListRecyclerView.adapter as ShoppingListAdapter).setItems(items)
+		val zeroItemsComparator = Comparator<RequiredItem> { a, b ->
+			if (a.amount > 0 && b.amount == 0) -1
+			else if (a.amount == 0 && b.amount > 0) 1
+			else 0
+		}
+		val orderComparator = when (this.settings.order) {
+			Order.Unordered -> Comparator { _, _ -> 0 }
+			Order.Alphabetical -> compareBy { it.name }
+			Order.AmountAscending -> compareBy { it.amount }
+			Order.AmountDescending -> compareBy<RequiredItem> { it.amount }.reversed()
```

Comment:
This should use `compareByDescending`. You also correctly noticed that we explicitly need to set the generic type of the comparator and I'd much prefer if it was on the first entry of the switch.

```suggestion
			Order.AmountDescending -> compareByDescending { it.amount }
```

Reply:
indeed, that's more readable.